shadowsocks/tcp.go

```diff
@@ -166,9 +161,29 @@ func proxyConnection(clientConn onet.DuplexConn, proxyMetrics *metrics.ProxyMetr
 	tgtConn := metrics.MeasureConn(tgtTCPConn, &proxyMetrics.ProxyTarget, &proxyMetrics.TargetProxy)
 
 	logger.Debugf("proxy %s <-> %s", clientConn.RemoteAddr().String(), tgtConn.RemoteAddr().String())
-	_, _, err = onet.Relay(clientConn, tgtConn)
-	if err != nil {
-		return onet.NewConnectionError("ERR_RELAY", "Failed to relay traffic", err)
+	fromClientErrCh := make(chan error)
+	go func() {
+		_, fromClientErr := clientConn.(io.WriterTo).WriteTo(tgtConn)
+		// Send FIN to target.
+		tgtConn.CloseWrite()
+		if fromClientErr != nil {
+			// Drain to prevent a close on cipher error.
+			clientConn.(io.WriterTo).WriteTo(ioutil.Discard)
```

Comment:
Won't this immediately return with an error on any subsequent decryption failure?  It seems like we would need to pierce the abstraction and drain the underlying `Conn`.

Reply:
Done